### PR TITLE
MGMT-2124 - Add cluster Id to API logs

### PR DESCRIPTION
**Pull request by slaviered**
/assign ronniel1
/assign gamli75

### CODE REVIEW

**Comment by slaviered** [comment]
Posted for initial CR for the concept. Not ready yet for merge

File: pkg/log/context.go [hunk 1]
```diff
@@ -36,3 +58,40 @@ func goid() string {
...
+func GetParamFromContext(ctx context.Context, key string) (string, string) {
+	val := ctx.Value(contextKey(paramKey(key)))
+	if val == nil {
+		return key, ""
+	} else {
+		return key, val.(string)
+	}
+}
+
+//openapi middleware handler that takes URL params from the path and put them
```
Comment (ronniel1) [inline pkg/log/context.go:75]: Comment is wrong.
This is taking specific params from the matched-route and placing in the context

File: pkg/log/context.go [hunk 2]
```diff
@@ -36,3 +58,40 @@ func goid() string {
...
+//on the context for further use (mainly by logs). parameters are prefixed to avoid
+//conflicts (for example: cluster_id --> PARAM$cluster_id)
+func ContextHandler() func(http.Handler) http.Handler {
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			//extract the Route from the request
+			mr := middleware.MatchedRouteFrom(r)
+			//set the parameters on the request's context
+			ctx := r.Context()
```
Comment (ronniel1) [inline pkg/log/context.go:85]: This should be done only if mr is not nil

File: pkg/log/context.go [hunk 3]
```diff
@@ -36,3 +58,40 @@ func goid() string {
 	idField := strings.Fields(strings.TrimPrefix(string(buf[:n]), "goroutine "))[0]
 	return idField
 }
+
+func paramKey(name string) string {
+	return PARAM + name
```
Comment (ronniel1) [inline pkg/log/context.go:63]: Why do we need to add a constant?

File: pkg/log/context.go [hunk 4]
```diff
@@ -36,3 +58,40 @@ func goid() string {
...
+func ContextHandler() func(http.Handler) http.Handler {
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			//extract the Route from the request
+			mr := middleware.MatchedRouteFrom(r)
+			//set the parameters on the request's context
+			ctx := r.Context()
+			if mr != nil {
+				for _, p := range mr.Params {
```
Comment (ronniel1) [inline pkg/log/context.go:87]: Why copy all the params to context?
We don't really need that.
Instead of iterating we can just check if the values we want exist and only copy them

**Review by ronniel1 — COMMENTED** [review]
Good job Sarah
Some comments

File: pkg/log/context.go (hunk 3 above)
Comment (slaviered) [reply pkg/log/context.go:63]: since I am adding all route parameters in the context (for generality purposes) I prefixed them to not create collision with other log fields across the application

File: pkg/log/context.go (hunk 4 above)
Comment (slaviered) [reply pkg/log/context.go:87]: I thought that since route params are important  (usually meaningful indices) its good idea to copy them all (there aren't that many) so if we want to add another index to the logs we would have it handy. and it saves a lot of if-then clauses

Comment (ronniel1) [reply pkg/log/context.go:93]: I don't understand what if-then clauses?
Why not just directly get the params we are interested in?

File: pkg/log/logcontext_test.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+	IsUser:   true,
+	Username: "dummy",
+}
+
+var authenticator = func(name string, _ string, authenticate security.TokenAuthentication) runtime.Authenticator {
+	return security.HttpAuthenticator(func(r *http.Request) (bool, interface{}, error) {
+		return true, &testUser, nil
+	})
+}
+
```
Comment (ronniel1) [inline pkg/log/logcontext_test.go:58]: Why is this needed in the context of the log test?
We are not testing authentication

File: pkg/log/logcontext_test.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+		server = createServer(logger, &installer)
+		bmclient = createClient()
+	})
+
+	AfterEach(func() {
+		server.Close()
+	})
+
+	Context("cluster_id", func() {
+		It("should be populated if there is such route param", func() {
```
Comment (ronniel1) [inline pkg/log/logcontext_test.go:126]: Also add check for host_id

**Review by ronniel1 — COMMENTED** [review]
nice

File: pkg/log/logcontext_test.go (hunk 5 above)
Comment (slaviered) [reply pkg/log/logcontext_test.go:58]: Since the flow go trough the server I need to disable authentication. However, since the tests are in the log package - which is a low level package - I could not import the auth package and disable authentication in more reasonable way. The solution was these callbacks that do the job. Either way there would have been some lines in the tests that meant to disable auth

File: pkg/log/context.go [hunk 7]
```diff
@@ -2,13 +2,23 @@ package log
...
 	"github.com/openshift/assisted-service/pkg/requestid"
 	"github.com/sirupsen/logrus"
 )
 
+type contextKey string
+
+//url parameters that should be added as fields to logs
+const (
+	PARAM          = "PARAM$" //prefix for all params
+	clusterIdParam = "cluster_id"
```
[the remark this answers is not in the document]
Comment (slaviered) [reply pkg/log/context.go:19]: DONE

File: pkg/log/context.go [hunk 8]
```diff
@@ -36,3 +58,40 @@ func goid() string {
 	idField := strings.Fields(strings.TrimPrefix(string(buf[:n]), "goroutine "))[0]
 	return idField
 }
+
+func paramKey(name string) string {
+	return PARAM + name
+}
+
+func GetParamFromContext(ctx context.Context, key string) (string, string) {
```
[the remark this answers is not in the document]
Comment (slaviered) [reply pkg/log/context.go]: DONE. (replaced with getParam)

File: pkg/log/context.go (hunk 2 above)
Comment (slaviered) [reply pkg/log/context.go:91]: DONE

File: pkg/log/logcontext_test.go [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+		EventsAPI:           nil,
+		Logger:              logger.Printf,
+		VersionsAPI:         nil,
+		ManagedDomainsAPI:   nil,
+		InnerMiddleware:     ContextHandler(),
+	})
+
+	server := &http.Server{Addr: "localhost:8082", Handler: h}
+	go serv(server)
+	time.Sleep(time.Second * 1) // Allow the server to start
```
Comment (filanov) [inline pkg/log/logcontext_test.go:68]: Always bad to add sleep in tests. it's better to create a polling function

File: pkg/log/logcontext_test.go [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+					FromContext(ctx, logger).Info("say something")
+					return installer.NewGetClusterOK()
+				}).Times(1)
+
+			_, err := bmclient.Installer.GetCluster(context.TODO(), &clientInstaller.GetClusterParams{
+				ClusterID: strfmt.UUID(uuid.New().String()),
+			})
+
+			Expect(err).NotTo(HaveOccurred())
+			Expect(logOut.String()).To(ContainSubstring("cluster_id="))
```
Comment (filanov) [inline pkg/log/logcontext_test.go:133]: Probably should validate the id as well. 
Same comment for all the tests.

File: pkg/log/logcontext_test.go [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+	var i int
+	for i = 0; i < 100 && err != nil; i++ {
+		_, err = bmclient.Installer.ListClusters(context.TODO(), &clientInstaller.ListClustersParams{})
+		time.Sleep(time.Millisecond * 10)
+	}
+
+	if err != nil {
+		panic("server took too long to start " + err.Error())
+	} else {
+		fmt.Printf("server is up after %d ms\n", i*10)
```
Comment (filanov) [inline pkg/log/logcontext_test.go:98]: It's a unit test we don't need those logs in it.

**Comment by filanov** [comment]
/hold
@romfreiman waiting for your decision if you want to delay after the release.

File: pkg/log/logcontext_test.go [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+func waitForServer(bmclient *client.AssistedInstall, mockInstallApi *mocks.MockInstallerAPI) {
+	var err error = fmt.Errorf("start polling server...")
+	//loop up to a second to wait for the server to go up
+	mockInstallApi.EXPECT().ListClusters(gomock.Any(), gomock.Any()).DoAndReturn(
+		func(ctx context.Context, params installer.ListClustersParams) middleware.Responder {
+			return installer.NewListClustersOK()
+		}).AnyTimes()
+	var i int
```
Comment (filanov) [inline pkg/log/logcontext_test.go:89]: you can probably replace all this block with https://godoc.org/github.com/onsi/gomega#Eventually

Comment (slaviered) [reply pkg/log/logcontext_test.go:89]: see my reply offline

**Review by ronniel1 — APPROVED** [review]
/lgtm
/approve
